### PR TITLE
Expose global variables to conans.tools only

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: Fix: Refactor code to limit scope of globals variables to `conans.tools`

- [x] Check pyinstaller regarding [this comment](https://github.com/conan-io/conan/pull/3636/files#diff-90eb62eb0f8a8aae8a4833ccd6bb523eR4).
- [x] Add changelog.

Rationale for this PR: 

We all don't feel comfortable having around global variables, but the use case in Conan is licit once you understand it (more below); so, as we have to live with them, let's try to limit their usage to a single module.

Why global variables in Conan? There is a very handy module from the user perspective to be used in the recipes, it is ``from conans import tools``, it is [documented](https://docs.conan.io/en/latest/reference/tools.html) and provides really useful stuff. But some of these tools require some components that may depend on the Conan configuration that the user has defined in the ``conan.conf`` file. So here is the problem: how to keep it simple for the user, but also executing those tools using the _global_ configuration, and here is where globals are needed. Output and URL requester are configured in the main Conan flow but then used in this handy tools without being passed in the arguments (and these tools are standalone functions).

Once we agree that globals are necessary, this PR tries to keep them inside the ``conans.tools`` module, no one else needs to use them, not even the main Conan workflow. This PR is a starting point and more work has to be done to achieve the final desired status:
- Main flow reads configuration files (or command line, or env variables) and creates an ``output`` and a ``requester`` objects.
- Main flow will carry around these objects call after call to reach the points where they are needed.
- Also, will set them in the globals variables, but just to be used by ``conans.tools``.

What do we get with this? Getting rid of globals variables make it easier for the developer, it's about removing collateral effects, unexpected behavior,... it also makes it easier to use Conan as a python library and to include it in others python applications.

Work to be done:
- There are several places marked with a warning in case the ``output`` or the ``requester`` are not passed as arguments, but instead of failing, the fallback behavior is to use the _global_ ones. This should be removed, and those arguments must be required. <-- affect only to calls from Conan itself, not to the module ``conans.tools`` that is exposed to the user.
- Remove from ``conans.tools`` what is not documented (at this moment it is exported _almost everything_).
- Work on tests, and refactor those that make use of the ``set_global_instances``. Tests cannot depend on a _global state_ if we want to run them in parallel.

As this PR affects to too many files, here are some links to the most relevant changes:
- Most of the files: replaced ``from conans import tools`` to ``form conans.client import tools`` in order to avoid exposition to global variables.
- [keep previous behavior](https://github.com/conan-io/conan/pull/3636/files#diff-f12516ecc39a5cc3e4b00de20632c010R228): if ``output`` is not given, fallback to ``_global_output``. This will happen only to calls from the Conan core, not from the user interface.
- [test the function](https://github.com/conan-io/conan/pull/3636/files#diff-b883adb5952830b5e83f71d66736f05dL69): tests do not need to modify globals in order to exercise a function.
- [globals only inside ``conans.tools``](https://github.com/conan-io/conan/pull/3636/files#diff-65dbc432e451dbebbbdb2dbf4a772533L32): global variables appears only in module ``conans.tools``.
- [currify ``output`` and ``requester``](https://github.com/conan-io/conan/pull/3636/files#diff-65dbc432e451dbebbbdb2dbf4a772533R82) when used from the module ``conans.tools``, those are the functions that the user will invoke from the conanfile.

----
TL; DR;

image

😄 😄 😄

### CODE REVIEW

**Comment by lasote** [comment]
My opinion: The users are not the culprit of the badly designed interfaces at the first time. I don't want to print warnings nor encourage to pass more parameters that doesn't make sense for them.

Study for Conan 2.0 a way to fix the "global variables situation" like instancing a "tools" object passing "self", as any other Conanfile collaborator.

**Comment by jgsogo** [comment]
The meme was that, just a meme... I keep the globals variables, I don't add parameters to the tools exposed to the users, warnings will never be prompted for these tools.

Warnings will be prompted if you (as developer of the Conan core) use some of this functions without providing an ``output`` or ``requester`` (where needed).

The aim of this PR is just to restrict the scope and usage of the global_variables to the tools exposed to the user.

File: conans/tools.py [hunk 1]
```diff
@@ -1,49 +1,211 @@
...
+    else:
+        files_save(path=path, content=content, only_if_modified=False)
+
+
+# From conans.client.tools.net
+ftp_download = tools_net.ftp_download
+
+
+def download(*args, **kwargs):
+    return tools_net.download(out=_global_output, requester=_global_requester, *args, **kwargs)
```
Comment (jgsogo) [inline conans/tools.py]: currify with ``partial``

File: conans/client/tools/system_pm.py [hunk 2]
```diff
@@ -121,107 +126,170 @@ def _install_any(self, packages):
...
 
 class NullTool(object):
+    def __init__(self, output=None):
+        if output is None:
+            import warnings
+            warnings.warn("Provide the output argument explicitly")
+            from conans.tools import _global_output
+            output = _global_output
+        self._output = output
+
```
Comment (jgsogo) [inline conans/client/tools/system_pm.py]: Do it once in a base class

**Comment by jgsogo** [comment]
@memsharded, I would like you to take a quick view at this PR, to know your opinion about the work done up to here so I continue with it, or abandon and close it. Thanks!

File: conans/model/conan_file.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import os
 from contextlib import contextmanager
 
-from conans import tools  # @UnusedImport KEEP THIS! Needed for pyinstaller to copy to exe.
+# TODO: from conans import tools  # @UnusedImport KEEP THIS! Needed for pyinstaller to copy to exe.
```
[the remark this answers is not in the document]
Comment (danimtb) [reply conans/model/conan_file.py]: Maybe this is needed for pyinstaller to "copy" the tools import to the bundle, otherwise people using the standalone installer might run into issues when using from `conans import tools` in their recipes. But not sure, will have to be tested

**Comment by danimtb** [comment]
I like the idea of this PR, I think it is better organized and makes sense to initialize the output and so on in the same place. Currently the import structure of the tools module is a bit of a mess, and this makes the hierarchy more clear.

File: conans/model/conan_file.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgsogo) [reply conans/model/conan_file.py]: I make this comment just to be sure that we check if this line is still required.

File: conans/test/util/tools_test.py [hunk 4]
```diff
@@ -32,19 +31,19 @@
...
-from conans.tools import OSInfo, SystemPackageTool, replace_in_file, AptTool, ChocolateyTool,\
-    set_global_instances
+from conans.client.tools.files import which, replace_in_file
+from conans.client.tools.system_pm import OSInfo, AptTool, ChocolateyTool, SystemPackageTool
+from conans.client import tools
 from conans.util.files import save, load, md5, mkdir
 import requests
 
 from nose.plugins.attrib import attr
+from conans.tools import os_info as global_os_info, get_global_instances
```
Comment (jgsogo) [inline conans/test/util/tools_test.py]: I don't want globals to be in tests, read comments below.

**Comment by lasote** [comment]
conflicts

**Comment by jgsogo** [comment]
conflicts solved! 😉

File: conans/client/tools/files.py [hunk 5]
```diff
@@ -63,8 +61,15 @@ def unzip(filename, destination=".", keep_permissions=False, pattern=None):
     that the zip was created correctly.
     :param pattern: Extract only paths matching the pattern. This should be a
     Unix shell-style wildcard, see fnmatch documentation for more details.
+    :param output: output
     :return:
     """
+    if output is None:
```
Comment (memsharded) [inline conans/client/tools/files.py]: This is highly repeated, move to function?

File: conans/model/conan_file.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (memsharded) [reply conans/model/conan_file.py]: It was required for sure. Be careful, because removing it, all tests pass, but it is later in the pyinstaller, which has missing modules.

File: conans/client/tools/files.py (hunk 5 above)
Comment (jgsogo) [reply conans/client/tools/files.py]: It is to be removed, I prefer to have it repeated and delete it one by one than writing a function that may be used by someone else and pervive, or may become dead code. But no strong opinion about this.

**Comment by jgsogo** [comment]
I keep the `if not output: output=global_output` because there can be recipes using _internal_ methods, and those will be broken if I remove the fallback snippet. I think it is safer to add the warning notice now and remove the _deprecated_ behavior in a breaking release (v2.0?).

**Comment by lasote** [comment]
conflicts

**Comment by jgsogo** [comment]
That is the reason why there are unsolved comments in the PR, anyway I should add a checkbox in the PR text too to make it clearer.

Besides that, I would really appreciate feedback/review, as there is still work to do and I need to know if we all agree with this approach.

**Comment by lasote** [comment]
I agree with the purpose and ok with the implementation.

File: conans/tools.py (hunk 1 above)
Comment (jgsogo) [reply conans/tools.py]: Not using `partial` because it will allow the user to override `out` and `requester` arguments (yes, partial allows to pass already _curryfied_ ones).

**Comment by jgsogo** [comment]
On the road to `develop`, final round of reviews!

Regarding the **pyinstaller** commented line, at least, in Mac, it works.

**Comment by lasote** [comment]
Confirmed also that the Debian installer is working and running in Linux correctly. I agree to remove that line. If any problem appears we will release a minor, but I think it is unlikely. 🍔

**Comment by lasote** [comment]
Windows working too. 🍔

File: conans/test/util/tools_test.py [hunk 6]
```diff
@@ -360,12 +359,12 @@ def __call__(self, command, *args, **kwargs):  # @UnusedVariable
             "CONAN_SYSREQUIRES_SUDO": "True"
         }):
             packages = ["verify_package", "verify_another_package", "verify_yet_another_package"]
-            runner = RunnerMultipleMock(["sudo -A apt-get update"])
-            spt = SystemPackageTool(runner=runner, tool=AptTool())
+            runner = RunnerMultipleMock(["sudo --askpass apt-get update"])
```
Comment (memsharded) [inline conans/test/util/tools_test.py]: Check if this is from another PR

File: conans/tools.py [hunk 7]
```diff
@@ -1,49 +1,210 @@
...
+
+
+def vcvars_command(*args, **kwargs):
+    return tools_win.vcvars_command(output=_global_output, *args, **kwargs)
+
+
+
+# Ready to use objects.
+try:
+    os_info = OSInfo()
```
Comment (memsharded) [inline conans/tools.py]: This seems not used along the PR. The access to ``os_info`` seems to have been replaced by creation of ``OSInfo()`` per user.

File: conans/test/util/tools_test.py (hunk 6 above)
Comment (jgsogo) [reply conans/test/util/tools_test.py]: After merging develop (with the revisions branch), there appeared new tests and modifications related to `--askpass`/`-A`.  I think it is not so important to have those mixed. Also, there is #3969, and eventually we will need to change all of them to `-A`. 

Nevertheless, if you think it is a problem, I can uniformize them all.

File: conans/tools.py (hunk 7 above)
Comment (jgsogo) [reply conans/tools.py]: ...but an existing recipe can be using it (https://docs.conan.io/en/latest/reference/conanfile/methods.html#system-requirements)

File: conans/client/tools/system_pm.py [hunk 8]
```diff
@@ -6,15 +6,20 @@
...
 from conans.util.env_reader import get_env
 
-_global_output = None
-
 
 class SystemPackageTool(object):
 
-    def __init__(self, runner=None, os_info=None, tool=None, recommends=False):
+    def __init__(self, runner=None, os_info=None, tool=None, recommends=False, output=None):
+        if output is None:
```
[the remark this answers is not in the document]
Comment (jgsogo) [reply conans/client/tools/system_pm.py]: Functionality moved to (new) `conans.util.fallback`
